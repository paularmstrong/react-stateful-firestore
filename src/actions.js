// @flow
import { getCollectionQueryPath, getQueryId, getDocumentIdsForQuery } from './modules/query';
import { createActionType, createRequestActionTypes } from './modules/actionTypes';

import type { Auth } from '@firebase/auth';
import type { Firestore, Query } from '@firebase/firestore';
import type { Reference } from '@firebase/storage';
import type { StoreState } from './reducers';
import type { FluxStandardAction } from './reducers/flux-standard-action';

type Action = FluxStandardAction<string, any, {}>;
type $ThunkAction<R> = (dispatch: Dispatch, getState: GetState, args: ThunkArgs) => R;
type ThunkAction = $ThunkAction<any>;
type Dispatch = (action: Action | ThunkAction) => any;
type GetState = () => StoreState;
type ThunkArgs = { auth: Auth, firestore: Firestore };

export const COLLECTIONS = {
  MODIFY: createActionType('collections/MODIFY'),
  MODIFY_ONE: createActionType('collections/MODIFY_ONE'),
  REMOVE: createActionType('collections/REMOVE')
};

export const QUERIES = {
  ...createRequestActionTypes('queries'),
  ADD: createActionType('queries/ADD'),
  REMOVE: createActionType('queries/REMOVE')
};

const onIdle = (fn: () => void) => (window.requestIdleCallback ? window.requestIdleCallback(fn) : setTimeout(fn, 1));

const _handleReceiveSnapshot = (dispatch: Dispatch, query, queryId) => (snapshot) => {
  const actions = [];
  if (snapshot.docChanges) {
    actions.push({ type: COLLECTIONS.MODIFY, payload: snapshot, meta: { query } });
  } else {
    actions.push({ type: COLLECTIONS.MODIFY_ONE, payload: snapshot, meta: { query } });
  }
  actions.push({ type: QUERIES.SUCCESS, payload: snapshot, meta: { queryId } });
  dispatch(actions);
};

export const addQuery = (query: Query, queryIdPrefix: string = '') => (
  dispatch: Dispatch,
  getState: GetState
): Promise<any> => {
  const state = getState();
  const { queries } = state;
  const queryId = `${queryIdPrefix}${getQueryId(query)}`;
  const collectionPath = `${queryIdPrefix}${getCollectionQueryPath(query)}`;

  const meta = { queryId };

  if (queryId !== collectionPath && collectionPath in queries) {
    const documentIds = getDocumentIdsForQuery(query, state);
    return Promise.resolve(dispatch({ type: QUERIES.ADD, payload: documentIds, meta }));
  }

  if (queryId in queries) {
    return Promise.resolve();
  }

  dispatch({ type: QUERIES.REQUEST, payload: query, meta });

  return query
    .get()
    .then(_handleReceiveSnapshot(dispatch, query, queryId))
    .catch((error) => {
      dispatch({ error: true, type: QUERIES.FAILURE, payload: error, meta });
    });
};

export const removeQuery = (query: Query, queryIdPrefix: string = '') => ({
  type: QUERIES.REMOVE,
  payload: query,
  meta: { queryId: `${queryIdPrefix}${getQueryId(query)}` }
});

export const LISTENERS = {
  ADD: createActionType('listeners/ADD'),
  REMOVE: createActionType('listeners/REMOVE')
};

export const addListener = (query: Query, queryIdPrefix: string = '') => (
  dispatch: Dispatch,
  getState: GetState
): Promise<any> => {
  const { listeners } = getState();
  const queryId = `${queryIdPrefix}${getQueryId(query)}`;
  const collectionPath = `${queryIdPrefix}${getCollectionQueryPath(query)}`;

  if (queryId in listeners || collectionPath in listeners) {
    return Promise.resolve();
  }

  return new Promise((resolve) => {
    onIdle(() => {
      const unsubscribe = query.onSnapshot(_handleReceiveSnapshot(dispatch, query, queryId));
      resolve(dispatch({ type: LISTENERS.ADD, payload: unsubscribe, meta: { queryId } }));
    });
  });
};

export const removeListener = (query: Query, queryIdPrefix: string = '') => (
  dispatch: Dispatch,
  getState: GetState
): Promise<any> => {
  const queryId = `${queryIdPrefix}${getQueryId(query)}`;
  const { listeners } = getState();

  if (listeners[queryId]) {
    listeners[queryId].unsubscribe();
    return Promise.resolve(
      dispatch({
        type: LISTENERS.REMOVE,
        meta: { queryId }
      })
    );
  }

  return Promise.resolve();
};

const _authQueryPrefix = 'auth|';
export const AUTH = {
  CHANGE: createActionType('auth/CHANGE')
};

export const setUser = (currentUser?: { uid: string }, userCollection?: string) => (
  dispatch: Dispatch,
  getState: GetState,
  { firestore }: ThunkArgs
): Promise<any> => {
  const actions = [];
  if (currentUser && userCollection) {
    const query = firestore.doc(`${userCollection}/${currentUser.uid}`);
    actions.push(addQuery(query, _authQueryPrefix));
    actions.push(addListener(query, _authQueryPrefix));
  }
  actions.push({ type: AUTH.CHANGE, payload: currentUser });
  return Promise.resolve(dispatch(actions));
};

export const unsetUser = (uid?: string, userCollection?: string) => (
  dispatch: Dispatch,
  getState: GetState,
  { firestore }: ThunkArgs
): Promise<any> => {
  const actions = [];
  if (uid) {
    if (userCollection) {
      const query = firestore.doc(`${userCollection}/${uid}`);
      actions.push(removeQuery(query, _authQueryPrefix));
      actions.push(removeListener(query, _authQueryPrefix));
      actions.push({ type: COLLECTIONS.REMOVE, payload: { id: uid }, meta: { query } });
    }
    actions.push({ type: AUTH.CHANGE });
    dispatch(actions);
  }
  return Promise.resolve();
};

export const STORAGE = {
  URL: createRequestActionTypes('storage/downloadUrl'),
  METADATA: createRequestActionTypes('storage/metadata')
};

export const getStorageDownloadUrl = (reference: Reference) => (
  dispatch: Dispatch,
  getState: GetState
): Promise<any> => {
  const meta = { reference };
  const state = getState();
  const ref = state.storage[reference.fullPath];
  if (ref && ref.downloadUrl) {
    return Promise.resolve(ref);
  }

  dispatch({ type: STORAGE.URL.REQUEST, meta });
  return reference
    .getDownloadURL()
    .then((url) => {
      return dispatch({ type: STORAGE.URL.SUCCESS, payload: url, meta });
    })
    .catch((error) => {
      dispatch({ error: true, type: STORAGE.URL.FAILURE, payload: error, meta });
      throw error;
    });
};

export const getStorageMetadata = (reference: Reference) => (dispatch: Dispatch, getState: GetState): Promise<any> => {
  const meta = { reference };
  const state = getState();
  const ref = state.storage[reference.fullPath];
  if (ref && ref.metadata) {
    return Promise.resolve(ref);
  }

  dispatch({ type: STORAGE.METADATA.REQUEST, meta });
  return reference
    .getMetadata()
    .then((data) => {
      return dispatch({ type: STORAGE.METADATA.SUCCESS, payload: data, meta });
    })
    .catch((error) => {
      dispatch({ error: true, type: STORAGE.METADATA.FAILURE, payload: error, meta });
      throw error;
    });
};
