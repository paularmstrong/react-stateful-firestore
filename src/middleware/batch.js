// @flow

import type { StoreState } from '../reducers';
import type { FluxStandardAction } from '../reducers/flux-standard-action';
import type { Reducer } from 'redux';

type Action = FluxStandardAction<string, any, {}>;
type Dispatch = (action: Action) => any;
type Next = (action: Action) => any;

const BATCH_ACTION = 'BATCH';

const isPlainAction = (action: { type: string } | string): boolean =>
  typeof action === 'object' && typeof action.type === 'string';

const actionBatch = (actions: Array<Action>) => {
  const meta = actions.reduce((obj, action) => {
    if (typeof action.meta === 'object') {
      Object.assign(obj, action.meta);
    }
    return obj;
  }, {});
  return {
    type: BATCH_ACTION,
    payload: actions,
    meta
  };
};

export const batchMiddleware = ({ dispatch }: { dispatch: Dispatch }) => (next: Next) => (
  actions: Array<Action> | Action
) => {
  if (Array.isArray(actions)) {
    const plainActions = [];
    actions.forEach((action) => {
      if (typeof action === 'function') {
        dispatch(action);
      } else if (isPlainAction(action)) {
        plainActions.push(action);
      }
    });
    return next(actionBatch(plainActions));
  } else {
    return next(actions);
  }
};

export const batchReducer = (reducer: Reducer<StoreState, Action>) => (state: StoreState, action: Action) => {
  return action && action.type === BATCH_ACTION ? action.payload.reduce(reducer, state) : reducer(state, action);
};
