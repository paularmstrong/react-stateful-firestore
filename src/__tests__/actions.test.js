import * as Actions from '../actions';
import { batchMiddleware } from '../middleware/batch';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const query = { id: 'foo', path: 'foo', get: () => Promise.resolve({}), onSnapshot: (cb) => cb({}) };
const firestore = {
  doc: (path) => ({ ...query, id: path, path })
};
const middlewares = [thunk.withExtraArgument({ firestore }), batchMiddleware];
const mockStore = configureStore(middlewares);
window.requestIdleCallback = jest.fn((cb) => cb());

const defaultState = { auth: {}, collections: {}, listeners: {}, queries: {}, storage: {} };
describe('Actions', () => {
  let store;
  beforeEach(() => {
    store = mockStore(defaultState);
  });

  describe('addQuery', () => {
    test('adds the query and executes it', () => {
      return store.dispatch(Actions.addQuery(query)).then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    test('does nothing if the query is already cached', () => {
      store = mockStore({ ...defaultState, queries: { foo: {} } });
      return store.dispatch(Actions.addQuery(query)).then(() => {
        expect(store.getActions()).toEqual([]);
      });
    });

    test('allows prefixing the queryId', () => {
      return store.dispatch(Actions.addQuery(query, 'tacos|')).then(() => {
        const queryIds = store
          .getActions()
          .filter((action) => !!action.meta.queryId)
          .map((action) => action.meta.queryId);
        expect(queryIds).toEqual(['tacos|foo', 'tacos|foo']);
      });
    });
  });

  describe('addListener', () => {
    test('starts the listener, then adds it', () => {
      return store.dispatch(Actions.addListener(query)).then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    test('does nothing if the listener is active', () => {
      store = mockStore({ ...defaultState, listeners: { foo: {} } });
      return store.dispatch(Actions.addListener(query)).then(() => {
        expect(store.getActions()).toEqual([]);
      });
    });

    test('allows prefixing the queryId', () => {
      return store.dispatch(Actions.addListener(query, 'tacos|')).then(() => {
        const queryIds = store
          .getActions()
          .filter((action) => !!action.meta.queryId)
          .map((action) => action.meta.queryId);
        expect(queryIds).toEqual(['tacos|foo', 'tacos|foo']);
      });
    });
  });

  describe('removeListener', () => {
    test('unsubscribes from the listener before removing', () => {
      const unsubscribe = jest.fn();
      const store = mockStore({ ...defaultState, listeners: { foo: { unsubscribe } } });
      return store.dispatch(Actions.removeListener(query)).then(() => {
        expect(unsubscribe).toHaveBeenCalled();
        expect(store.getActions()).toEqual([{ type: Actions.LISTENERS.REMOVE, meta: { queryId: 'foo' } }]);
      });
    });
  });

  describe('setUser', () => {
    test('adds queries and listeners', () => {
      return store.dispatch(Actions.setUser({ uid: '123' }, 'users')).then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    test('triggers only auth change action if no collection', () => {
      return store.dispatch(Actions.setUser({ uid: '123' })).then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
    });
  });

  describe('unsetUser', () => {
    test('removes queries, listeners, and the user from the collection', () => {
      store = mockStore({ ...defaultState, listeners: { 'auth|users/123': { unsubscribe: jest.fn() } } });
      return store.dispatch(Actions.unsetUser('123', 'users')).then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    test('triggers only auth change action if no collection', () => {
      store = mockStore(defaultState);
      return store.dispatch(Actions.unsetUser('123')).then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
    });
  });

  describe('getStorageDownloadUrl', () => {
    test('requests download URL', () => {
      store = mockStore(defaultState);
      const getDownloadURL = jest.fn(() => Promise.resolve());
      return store.dispatch(Actions.getStorageDownloadUrl({ fullPath: '/thing', getDownloadURL })).then(() => {
        expect(store.getActions()).toMatchSnapshot();
        expect(getDownloadURL).toHaveBeenCalled();
      });
    });

    test('does nothing if already in state', () => {
      store = mockStore({ ...defaultState, storage: { '/thing': { downloadUrl: '/thing' } } });
      const getDownloadURL = jest.fn(() => Promise.resolve());
      return store.dispatch(Actions.getStorageDownloadUrl({ fullPath: '/thing', getDownloadURL })).then(() => {
        expect(store.getActions()).toMatchSnapshot();
        expect(getDownloadURL).not.toHaveBeenCalled();
      });
    });
  });

  describe('getStorageMetadata', () => {
    test('requests metadata', () => {
      store = mockStore(defaultState);
      const getMetadata = jest.fn(() => Promise.resolve());
      return store.dispatch(Actions.getStorageMetadata({ fullPath: '/thing', getMetadata })).then(() => {
        expect(store.getActions()).toMatchSnapshot();
        expect(getMetadata).toHaveBeenCalled();
      });
    });

    test('does nothing if already in state', () => {
      store = mockStore({ ...defaultState, storage: { '/thing': { downloadUrl: '/thing', metadata: {} } } });
      const getMetadata = jest.fn(() => Promise.resolve());
      return store.dispatch(Actions.getStorageMetadata({ fullPath: '/thing', getMetadata })).then(() => {
        expect(store.getActions()).toMatchSnapshot();
        expect(getMetadata).not.toHaveBeenCalled();
      });
    });
  });
});
