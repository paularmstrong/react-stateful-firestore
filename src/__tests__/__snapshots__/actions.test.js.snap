// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions addListener starts the listener, then adds it 1`] = `
Array [
  Object {
    "meta": Object {
      "query": Object {
        "get": [Function],
        "id": "foo",
        "onSnapshot": [Function],
        "path": "foo",
      },
      "queryId": "foo",
    },
    "payload": Array [
      Object {
        "meta": Object {
          "query": Object {
            "get": [Function],
            "id": "foo",
            "onSnapshot": [Function],
            "path": "foo",
          },
        },
        "payload": Object {},
        "type": "firestore/collections/MODIFY_ONE",
      },
      Object {
        "meta": Object {
          "queryId": "foo",
        },
        "payload": Object {},
        "type": "firestore/queries/SUCCESS",
      },
    ],
    "type": "BATCH",
  },
  Object {
    "meta": Object {
      "queryId": "foo",
    },
    "payload": undefined,
    "type": "firestore/listeners/ADD",
  },
]
`;

exports[`Actions addQuery adds the query and executes it 1`] = `
Array [
  Object {
    "meta": Object {
      "queryId": "foo",
    },
    "payload": Object {
      "get": [Function],
      "id": "foo",
      "onSnapshot": [Function],
      "path": "foo",
    },
    "type": "firestore/queries/REQUEST",
  },
  Object {
    "meta": Object {
      "query": Object {
        "get": [Function],
        "id": "foo",
        "onSnapshot": [Function],
        "path": "foo",
      },
      "queryId": "foo",
    },
    "payload": Array [
      Object {
        "meta": Object {
          "query": Object {
            "get": [Function],
            "id": "foo",
            "onSnapshot": [Function],
            "path": "foo",
          },
        },
        "payload": Object {},
        "type": "firestore/collections/MODIFY_ONE",
      },
      Object {
        "meta": Object {
          "queryId": "foo",
        },
        "payload": Object {},
        "type": "firestore/queries/SUCCESS",
      },
    ],
    "type": "BATCH",
  },
]
`;

exports[`Actions getStorageDownloadUrl does nothing if already in state 1`] = `Array []`;

exports[`Actions getStorageDownloadUrl requests download URL 1`] = `
Array [
  Object {
    "meta": Object {
      "reference": Object {
        "fullPath": "/thing",
        "getDownloadURL": [MockFunction] {
          "calls": Array [
            Array [],
          ],
        },
      },
    },
    "type": "firestore/storage/downloadUrl/REQUEST",
  },
  Object {
    "meta": Object {
      "reference": Object {
        "fullPath": "/thing",
        "getDownloadURL": [MockFunction] {
          "calls": Array [
            Array [],
          ],
        },
      },
    },
    "payload": undefined,
    "type": "firestore/storage/downloadUrl/SUCCESS",
  },
]
`;

exports[`Actions getStorageMetadata does nothing if already in state 1`] = `Array []`;

exports[`Actions getStorageMetadata requests metadata 1`] = `
Array [
  Object {
    "meta": Object {
      "reference": Object {
        "fullPath": "/thing",
        "getMetadata": [MockFunction] {
          "calls": Array [
            Array [],
          ],
        },
      },
    },
    "type": "firestore/storage/metadata/REQUEST",
  },
  Object {
    "meta": Object {
      "reference": Object {
        "fullPath": "/thing",
        "getMetadata": [MockFunction] {
          "calls": Array [
            Array [],
          ],
        },
      },
    },
    "payload": undefined,
    "type": "firestore/storage/metadata/SUCCESS",
  },
]
`;

exports[`Actions setUser adds queries and listeners 1`] = `
Array [
  Object {
    "meta": Object {
      "queryId": "auth|users/123",
    },
    "payload": Object {
      "get": [Function],
      "id": "users/123",
      "onSnapshot": [Function],
      "path": "users/123",
    },
    "type": "firestore/queries/REQUEST",
  },
  Object {
    "meta": Object {
      "query": Object {
        "get": [Function],
        "id": "users/123",
        "onSnapshot": [Function],
        "path": "users/123",
      },
      "queryId": "auth|users/123",
    },
    "payload": Array [
      Object {
        "meta": Object {
          "query": Object {
            "get": [Function],
            "id": "users/123",
            "onSnapshot": [Function],
            "path": "users/123",
          },
        },
        "payload": Object {},
        "type": "firestore/collections/MODIFY_ONE",
      },
      Object {
        "meta": Object {
          "queryId": "auth|users/123",
        },
        "payload": Object {},
        "type": "firestore/queries/SUCCESS",
      },
    ],
    "type": "BATCH",
  },
  Object {
    "meta": Object {
      "queryId": "auth|users/123",
    },
    "payload": undefined,
    "type": "firestore/listeners/ADD",
  },
  Object {
    "meta": Object {},
    "payload": Array [
      Object {
        "payload": Object {
          "uid": "123",
        },
        "type": "firestore/auth/CHANGE",
      },
    ],
    "type": "BATCH",
  },
  Object {
    "meta": Object {
      "query": Object {
        "get": [Function],
        "id": "users/123",
        "onSnapshot": [Function],
        "path": "users/123",
      },
      "queryId": "auth|users/123",
    },
    "payload": Array [
      Object {
        "meta": Object {
          "query": Object {
            "get": [Function],
            "id": "users/123",
            "onSnapshot": [Function],
            "path": "users/123",
          },
        },
        "payload": Object {},
        "type": "firestore/collections/MODIFY_ONE",
      },
      Object {
        "meta": Object {
          "queryId": "auth|users/123",
        },
        "payload": Object {},
        "type": "firestore/queries/SUCCESS",
      },
    ],
    "type": "BATCH",
  },
]
`;

exports[`Actions setUser triggers only auth change action if no collection 1`] = `
Array [
  Object {
    "meta": Object {},
    "payload": Array [
      Object {
        "payload": Object {
          "uid": "123",
        },
        "type": "firestore/auth/CHANGE",
      },
    ],
    "type": "BATCH",
  },
]
`;

exports[`Actions unsetUser removes queries, listeners, and the user from the collection 1`] = `
Array [
  Object {
    "meta": Object {
      "queryId": "auth|users/123",
    },
    "type": "firestore/listeners/REMOVE",
  },
  Object {
    "meta": Object {
      "query": Object {
        "get": [Function],
        "id": "users/123",
        "onSnapshot": [Function],
        "path": "users/123",
      },
      "queryId": "auth|users/123",
    },
    "payload": Array [
      Object {
        "meta": Object {
          "queryId": "auth|users/123",
        },
        "payload": Object {
          "get": [Function],
          "id": "users/123",
          "onSnapshot": [Function],
          "path": "users/123",
        },
        "type": "firestore/queries/REMOVE",
      },
      Object {
        "meta": Object {
          "query": Object {
            "get": [Function],
            "id": "users/123",
            "onSnapshot": [Function],
            "path": "users/123",
          },
        },
        "payload": Object {
          "id": "123",
        },
        "type": "firestore/collections/REMOVE",
      },
      Object {
        "type": "firestore/auth/CHANGE",
      },
    ],
    "type": "BATCH",
  },
]
`;

exports[`Actions unsetUser triggers only auth change action if no collection 1`] = `
Array [
  Object {
    "meta": Object {},
    "payload": Array [
      Object {
        "type": "firestore/auth/CHANGE",
      },
    ],
    "type": "BATCH",
  },
]
`;
